### PR TITLE
https://www.pivotaltracker.com/story/show/149985390

### PR DESCRIPTION
**Pull request by kendelljoseph**
This is headless for browser testing.
- This will run a chrome/firefox browser that has no interface but is able to visit the site
- We can use this "headless browser" to see if elements are present, click things, etc.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,3 +1,9 @@
...
 node_js:
-  - "8.1.4"
\ No newline at end of file
+  - "8.1.4"
+before_install:
+  - wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb
+  - sudo dpkg -i google-chrome-stable_current_amd64.deb
+  - sudo apt-get install fonts-liberation libappindicator1 xdg-utils
+  - sudo apt-get install -f
+  - sudo apt-get install firefox xvfb
```
Comment (kendelljoseph) [inline .travis.yml:9]: These are the commands you would have to run in order to get the headless browser running.
- this means that these tests probably only pass on a linux machine.
- the cloud 9 space is a linux machine, and is setup to run these tests.
- this travis.yml is for continuous integration support (whenever we add it) through [travis](travis-ci.com)

File: features/step_definitions/client.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      .forBrowser(`chrome`)
+      .build();
+  });
+  
+  When(/^a client browser should shut down$/, function(){
+    return this.driver.quit();
+  });
+  
+  When(/^a client should be able to visit (.*)$/, function(path){
+    return this.driver.get(`localhost:${this.serverData.options.port}${path}`);
```
Comment (kendelljoseph) [inline features/step_definitions/client.js:26]: This is how you visit a site (any site really, but this one visits the locally running server)

File: features/step_definitions/client.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  });
+
+  When(/^a client should be able to see an element tagged (.*)$/, function(tag){
+    return this.driver.findElement(this.webdriver.By.css(`${tag}`));
```
Comment (kendelljoseph) [inline features/step_definitions/client.js:30]: After visiting a site, this is how you can select an element to see if it exists.

File: features/test/client/artprize.feature [hunk 4]
```diff
@@ -0,0 +1,10 @@
+Feature: Artprize demo
+  
+  Scenario: A person wants to visit the artprize demo
+    Given the server should start
+      And a client browser should start up
+      And a client should be able to visit /artprize
+      And a client should be able to see an element tagged [ng-app="blockforge"]
```
Comment (kendelljoseph) [inline features/test/client/artprize.feature:7]: Every demo will have its own set of tests.
This one is specifically checking the ArtPrize demo.

File: package.json [hunk 5]
```diff
@@ -4,7 +4,7 @@
   "description": "A lego block programming language",
   "main": "run_server.js",
   "scripts": {
-    "test": "./node_modules/.bin/cucumber.js",
+    "test": "xvfb-run ./node_modules/.bin/cucumber.js",
```
Comment (kendelljoseph) [inline package.json:7]: Tests now require the `xvfb-run` command before running. (for frontend testing).
- It is installed in the cloud 9 workspace and has been aliased to just `cucumber`
- so you can run tests using just `cucumber`  or `cucumber /features/test/testdir/testname.feature`

**Review by hdewey — APPROVED** [review]

**Pull request merged**